### PR TITLE
KAFKA-2733: Standardize metric name for Kafka Streams

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
Ping @ymatsuda @junrao for reviews. General summary:

Top-level metrics with Kafka Streams are completely owned by Kafka Streams framework itself, and is not exposed to the users. They have the following format:

1) groupName = "streaming-metrics": e.g. in JMX it will be used as "type=streaming-metrics" in mbean name.

2) tags = ["client-id" => clientId + "-" + getName()], hence the tags client-id is distinguished per StreamThread.

3) metricName = "poll-time-avg", "process-time-max", "commit-calls-rate", etc.

A single thread containing multiple tasks will use a single set of sensors, and there is no differentiating across tasks within the thread.

---

Second-level metrics is exposed through the ProcessorContext.metric() API, in which users can add their own metrics. Today we only contain state-store operation latency metrics, with the following format:

1) groupName = "streaming-" + scopeName + "-metrics", where scopeName defines the store type, e.g. "in-memory-state", "in-memory-lru-state", "rocksdb-state".

2) no additional tags from the framework, i.e. still have one tag as clientId, while we have APIs for users to add more tags if they want while developing their own state implementations.

3) we define metric name based on the operation name and the "entity" name (i.e. store name for this set of metrics) , in two levels:

A global metric with metricName e.g. "all-put", "all-get", "all-delete";
A per-store metric with metricName e.g. "store1-put", "store2-put", .. etc where the global metrics' sensor is the parent of the per-store metrics' sensor.

With this, we still only define metrics per-thread, not per-task within the thread, but distinguish different state-store instances.

**Comment by ymatsuda** [comment]
lgtm

**Comment by junrao** [comment]
To avoid merge conflict, you probably want to wait for https://github.com/apache/kafka/pull/328 to get in first.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -721,23 +721,25 @@ public Sensor addLatencySensor(String scopeName, String entityName, String opera
             for (int i = 0; i < tags.length; i += 2)
                 tagMap.put(tags[i], tags[i + 1]);
 
+            String metricGroupName = "streaming-" + scopeName + "-metrics";
```
Comment (junrao) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:724]: Would it be better to use scopeName as a tag when creating MetricName?

Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:724]: I have thought about it before, the reason I would prefer it to end with the groupName is to distinguish the top-level metrics (with groupName "streaming-metrics") that owned by the framework with the second-level metrics (with groupName "streaming-scope-metrics") that can be created by the users. And if scope is already part of the groupName, it is not necessary to be in the tags any more.

**Comment by guozhangwang** [comment]
@junrao Ping for another look. Jenkins failures seem not relevant and we have a separate ticket tracking it.

**Comment by junrao** [comment]
Thanks for the patch. LGTM

**Comment by guozhangwang** [comment]
Merged to trunk.

**Pull request closed without merging**